Mathlib/AlgebraicTopology/SimplexCategory/GeneratorsRelations/EpiMono.lean

```diff
@@ -98,4 +96,121 @@ lemma eq_or_len_le_of_P_δ {x y : SimplexCategoryGenRel} {f : x ⟶ y} (h_δ : P
+    obtain ⟨i, rfl⟩ := Fin.eq_succ_of_ne_zero (Fin.ne_zero_of_lt h)
+    rw [← Fin.le_castSucc_iff] at h
+    exact Or.inr ⟨i, i', by rw [δ_comp_σ_of_le h]⟩
+
+/-- A low-dimensional special case of the previous -/
+private lemma switch_δ_σ₀ (i : Fin 1) (i' : Fin 2) :
+    δ i' ≫ σ i = 𝟙 _ := by
+  fin_cases i; fin_cases i'
+  · exact δ_comp_σ_self
+  · exact δ_comp_σ_succ
+
+private lemma factor_δ_σ {n : ℕ} (i : Fin (n + 1)) (i' : Fin (n + 2)) :
+    ∃ (z : SimplexCategoryGenRel) (e : mk n ⟶ z) (m : z ⟶ mk n)
+      (_ : P_σ e) (_ : P_δ m), δ i' ≫ σ i = e ≫ m := by
+  cases n with
+  | zero => exact ⟨_, _, _, P_σ.id_mem _, P_δ.id_mem _, by simp [switch_δ_σ₀]⟩
+  | succ n =>
+    obtain h | ⟨j, j', h⟩ := switch_δ_σ i i'
+    · exact ⟨_, _, _, P_σ.id_mem _, P_δ.id_mem _, by simp [h]⟩
+    · exact ⟨_, _, _, P_σ.σ _, P_δ.δ _, h⟩
+
+/-- An auxiliary lemma that shows there exists a factorization as a P_δ followed by a P_σ for
+morphisms of the form `P_δ ≫ σ`. -/
+private lemma factor_P_δ_σ {n : ℕ} (i : Fin (n + 1)) {x : SimplexCategoryGenRel}
+    (f : x ⟶ mk (n + 1)) (hf : P_δ f) : ∃ (z : SimplexCategoryGenRel) (e : x ⟶ z) (m : z ⟶ mk n)
+      (_ : P_σ e) (_ : P_δ m), f ≫ σ i = e ≫ m := by
+  induction n generalizing x with
+  | zero => cases hf with
+    | of _ h => cases h; exact factor_δ_σ _ _
+    | id  => exact ⟨_, _, _, P_σ.σ i, P_δ.id_mem _, by simp⟩
+    | comp_of j f hf hg =>
+      obtain ⟨k⟩ := hg
+      obtain ⟨rfl, rfl⟩ | hf' := eq_or_len_le_of_P_δ hf
+      · simpa using factor_δ_σ i k
+      · simp at hf'
+  | succ n hn =>
+    cases hf with
+    | of _ h => cases h; exact factor_δ_σ _ _
+    | id n => exact ⟨_, _, _, P_σ.σ i, P_δ.id_mem _, by simp⟩
+    | comp_of f g hf hg =>
+      obtain ⟨k⟩ := hg
+      obtain ⟨rfl, rfl⟩ | h' := eq_or_len_le_of_P_δ hf
+      · simpa using factor_δ_σ i k
+      · obtain h'' | ⟨j, j', h''⟩ := switch_δ_σ i k
+        · exact ⟨_, _, _, P_σ.id_mem _, hf, by simp [h'']⟩
+        · obtain ⟨z, e, m, he, hm, fac⟩ := hn j f hf
+          exact ⟨z, e, m ≫ δ j', he, P_δ.comp_mem _ _ hm (P_δ.δ j'),
+            by simp [h'', reassoc_of% fac]⟩
+
+/-- Any morphism in `SimplexCategoryGenRel` can be decomposed as a `P_σ` followed by a `P_δ`. -/
+theorem exists_P_σ_P_δ_factorization {x y : SimplexCategoryGenRel} (f : x ⟶ y) :
+    ∃ (z : SimplexCategoryGenRel) (e : x ⟶ z) (m : z ⟶ y)
+        (_ : P_σ e) (_ : P_δ m), f = e ≫ m := by
+  induction f with
+  | @id n => use (mk n), (𝟙 (mk n)), (𝟙 (mk n)), P_σ.id_mem _, P_δ.id_mem _; simp
+  | @comp_δ n n' f j h =>
+    obtain ⟨z, e, m, ⟨he, hm, h⟩⟩ := h
+    rw [h, Category.assoc]
+    use z, e, m ≫ δ j, he, P_δ.comp_mem _ _ hm (P_δ.δ _)
+  | @comp_σ n n' f j h =>
+    obtain ⟨z, e, m, ⟨he, hm, h⟩⟩ := h
+    rw [h]
+    cases hm with
+    | of g hg =>
+      cases hg
+      rw [Category.assoc]
+      obtain ⟨z₁, e₁, m₁, ⟨he₁, hm₁, h₁⟩⟩ := factor_δ_σ j _
+      rw [h₁]
+      use z₁, e ≫ e₁, m₁, P_σ.comp_mem _ _ he he₁, hm₁
+      simp
```

Comment:
Could you improve the proofs similarly as above? Here, it is likely that all the `rw` can be put together in the `simp` call.